File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mid(ListNode head){
        ListNode slow = head ; 
        ListNode fast = head ; 
        while(fast!=null&&fast.next!=null){
            slow = slow.next ; 
            fast = fast.next.next ;
        }
        return slow ;
    }
    public ListNode reverse(ListNode head){
       ListNode prev = null ; 
       ListNode curr = head ;
       ListNode next = curr.next ;
       while(curr!=null){
            curr.next = prev ; 
            prev = curr ; 
            curr = next ;
            if(next!=null)
            next = next.next ;
       }
       return prev ;

    }
    public void reorderList(ListNode head) {
        if(head.next==null)
        return ;
        ListNode mid = mid(head) ;
        
        ListNode rev = reverse(mid) ;
        ListNode temp = head ; 
        ListNode prev = temp ; 
        ListNode next = temp.next ;
        ListNode x = null ;
        while(temp!=mid){
            temp.next = rev ; 
            x = rev.next ;
            rev.next = next ;
            prev = temp;
            temp = next ;
            rev = x ;
            next = next.next ;
        }
        if(temp!=null){
            temp.next = null ;
        }
        
            

    }
}

File: 0143-reorder-list/0143-reorder-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mid(struct ListNode* head) {
    struct ListNode* slow = head;
    struct ListNode* fast = head;
    while (fast != NULL && fast->next != NULL) {
        slow = slow->next;
        fast = fast->next->next;
    }
    return slow;
}

struct ListNode* reverse(struct ListNode* head) {
    struct ListNode* prev = NULL;
    struct ListNode* curr = head;
    struct ListNode* next = (curr != NULL) ? curr->next : NULL;
    while (curr != NULL) {
        curr->next = prev;
        prev = curr;
        curr = next;
        if (next != NULL) {
            next = next->next;
        }
    }
    return prev;
}

void reorderList(struct ListNode* head) {
    if (head->next == NULL) {
        return;
    }
    struct ListNode* midNode = mid(head);
    struct ListNode* rev = reverse(midNode);
    struct ListNode* temp = head;
    struct ListNode* prev = temp;
    struct ListNode* next = temp->next;
    struct ListNode* x = NULL;
    while (temp != midNode) {
        temp->next = rev;
        x = rev->next;
        rev->next = next;
        prev = temp;
        temp = next;
        rev = x;
        next = (next != NULL) ? next->next : NULL;
    }
    if (temp != NULL) {
        temp->next = NULL;
    }
}

